lib/vtls/schannel.c

```diff
@@ -405,6 +405,362 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
+
+#ifdef HAS_CLIENT_CERT_PATH
+  /* client certificate */
+  if(data->set.ssl.primary.clientcert) {
+    DWORD cert_store_name;
+    TCHAR *cert_store_path = NULL;
+    TCHAR *cert_thumbprint_str;
+    CRYPT_HASH_BLOB cert_thumbprint;
+    BYTE cert_thumbprint_data[CERT_THUMBPRINT_DATA_LEN];
+    HCERTSTORE cert_store;
+    FILE *fInCert = NULL;
+
+    TCHAR *cert_path = curlx_convert_UTF8_to_tchar(
+                           data->set.ssl.primary.clientcert);
+    if(!cert_path)
+      return CURLE_OUT_OF_MEMORY;
+
+    result = get_cert_location(cert_path, &cert_store_name,
+                               &cert_store_path, &cert_thumbprint_str);
+    if((result != CURLE_OK) && (data->set.ssl.primary.clientcert[0]!='\0'))
+      fInCert = fopen(data->set.ssl.primary.clientcert, "rb");
+
+    if((result != CURLE_OK) && !fInCert) {
+      failf(data, "schannel: Failed to get certificate location"
+            " or file for %s",
+            data->set.ssl.primary.clientcert);
+      curlx_unicodefree(cert_path);
+      return result;
+    }
+
+    if(fInCert) {
+      /* Reading a .P12 or .pfx file, like the example at bottom of
+         https://social.msdn.microsoft.com/Forums/windowsdesktop/
+         en-US/3e7bc95f-b21a-4bcd-bd2c-7f996718cae5
+      */
+      void *certdata = NULL;
+      long filesize = 0;
+      CRYPT_DATA_BLOB datablob;
+      WCHAR* pszPassword;
+      size_t pwd_len = 0;
+      int str_w_len = 0;
+      int continue_reading = fseek(fInCert, 0, SEEK_END) == 0;
+      if(continue_reading)
+        filesize = ftell(fInCert);
+      if(filesize < 0)
+        continue_reading = 0;
+      if(continue_reading)
+        continue_reading = fseek(fInCert, 0, SEEK_SET) == 0;
+      if(continue_reading)
+        certdata = malloc(((size_t)filesize) + 1);
+      if(!certdata ||
+         ((int) fread(certdata, (size_t)filesize, 1, fInCert) != 1))
+        continue_reading = 0;
+      fclose(fInCert);
+      curlx_unicodefree(cert_path);
+
+      if(!continue_reading) {
+        failf(data, "schannel: Failed to read cert file %s",
+              data->set.ssl.primary.clientcert);
+        free(certdata);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+
+      /* Convert key-pair data to the in-memory certificate store */
+      datablob.pbData = (BYTE*)certdata;
+      datablob.cbData = (DWORD)filesize;
+
+      if(data->set.ssl.key_passwd != NULL)
+        pwd_len = strlen(data->set.ssl.key_passwd);
+      pszPassword = (WCHAR*)malloc(sizeof(WCHAR)*(pwd_len + 1));
+      if(pwd_len > 0)
+        str_w_len =
+          MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS,
+                              data->set.ssl.key_passwd, (int)pwd_len,
+                              pszPassword, (int)(pwd_len + 1));
+
+      if((str_w_len >= 0) && (str_w_len <= (int)pwd_len))
+        pszPassword[str_w_len] = 0;
+      else
+        pszPassword[0] = 0;
+
+      cert_store = PFXImportCertStore(&datablob, pszPassword, 0);
+      free(pszPassword);
```

Comment:
Please overwrite the password with zeros first before releasing it. Otherwise other processes might obtain the cleartext password

Reply:
Existing code. Just pulled out into its own method. We can refactor later. Let's not distract ourselves.